python/ray/tests/test_modin.py

```diff
@@ -0,0 +1,402 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations under
+# the License.
+#
+# This file is copied and adapted from:
+# http://github.com/modin-project/modin/master/modin/pandas/test/test_general.py
+
+import sys
+import pytest
+import pandas
+import numpy as np
+from numpy.testing import assert_array_equal
+import ray
+from ray.util.client.ray_client_helpers import ray_start_client_server
+
+modin_compatible_version = sys.version_info >= (3, 7, 0)
+
+# These tests are written for versions of Modin that require python 3.7+
+pytestmark = pytest.mark.skipif(
+    not modin_compatible_version,
+    reason="Newer versions of Modin require Python 3.7+")
+
+if modin_compatible_version:
+    from ray.tests.modin_test_utils import df_equals
+    import modin.pandas as pd
+
+
+# Module scoped fixture. Will first run all tests without ray
+# client, then rerun all tests with a single ray client session.
+@pytest.fixture(params=[False, True], autouse=True, scope="module")
+def run_ray_client(request):
+    if request.param:
+        with ray_start_client_server() as client:
+            yield client
+    else:
+        # Run without ray client (do nothing)
+        yield
+        # Cleanup state before rerunning tests with client
+        ray.shutdown()
+
+
+random_state = np.random.RandomState(seed=42)
+
+# Size of test dataframes
+NCOLS, NROWS = (2**6, 2**8)
+
+# Range for values for test data
+RAND_LOW = 0
+RAND_HIGH = 100
+
+# Input data and functions for the tests
+# The test data that we will test our code against
+test_data = {
+    "int_data": {
+        "col{}".format(int((i - NCOLS / 2) % NCOLS + 1)): random_state.randint(
+            RAND_LOW, RAND_HIGH, size=(NROWS))
+        for i in range(NCOLS)
+    },
+    "float_nan_data": {
+        "col{}".format(int((i - NCOLS / 2) % NCOLS + 1)): [
+            x if (j % 4 == 0 and i > NCOLS // 2)
+            or (j != i and i <= NCOLS // 2) else np.NaN for j, x in enumerate(
+                random_state.uniform(RAND_LOW, RAND_HIGH, size=(NROWS)))
+        ]
+        for i in range(NCOLS)
+    },
+}
+
+test_data["int_data"]["index"] = test_data["int_data"].pop("col{}".format(
+    int(NCOLS / 2)))
+
+for col in test_data["float_nan_data"]:
+    for row in range(NROWS // 2):
+        if row % 16 == 0:
+            test_data["float_nan_data"][col][row] = np.NaN
+
+test_data_values = list(test_data.values())
+test_data_keys = list(test_data.keys())
+
+
+@pytest.mark.parametrize("data", test_data_values, ids=test_data_keys)
+def test_isna(data):
+    pandas_df = pandas.DataFrame(data)
+    modin_df = pd.DataFrame(data)
```

Comment:
Just making sure, modin is using Ray Client here and is not initializing another ray instance right?

Reply:
Yup, modin won't initialize a new instance if ray.is_initialized is already true